### PR TITLE
Remove xarray dependency 

### PR DESCRIPTION
**Pull request by stephaniegott**
Removes `xarray` from the dependencies of ipyleaflet.

Although ipyleaflet has `xarray` listed in its `install_requires` I'm not exactly sure why it needs this dependency. It would be great if when I installed ipyleaflet, I didn't also have to install `xarray` (which, in turn, pulls in `pandas`) as a transitive dependency.

### CODE REVIEW

**Comment by davidbrochart** [comment]
`xarray` is needed for the Velocity layer. With some work, we could probably do with just `numpy`. Or not include `xarray` in the requirements, and throw an error saying that `xarray` is needed if a Velocity layer is used.

**Comment by martinRenou** [comment]
> With some work, we could probably do with just `numpy`

I'd be in favor with this :+1:. In the end the Velocity layer only needs a grid of velocity data, which can definitely be expressed by a NumPy array.

This would:
- make the Velocity layer more general and simple (as people will be able to pass a NumPy array built from any kind of data structure: Pandas, xarray). And it would no longer be required to pass an xarray that has a very specific structure.
- remove the xarray dependency which, if I remember correctly, is quite heavy
- remove some code complexity from ipyleaflet

**Comment by stephaniegott** [comment]
@davidbrochart @martinRenou Thanks for the feedback. Subbing in `numpy` for `xarray` sounds like a good plan of action. In the meantime, could we perhaps add an `extras_require` for Velocity? Something like:
```
setup_args = {
    'name':'ipyleaflet',
    ...
    'extras_require': {
        'velocity':  ['xarray>=0.10'],
    }
)
```

**Comment by davidbrochart** [comment]
I'm not sure it is such a good idea to try and implement ourselves what xarray offers. Dimension-based indexing is really powerful. Say you only want to show a subset of the data in ipyleaflet, you can select the lat/lon range xarray. If it's big data, it can be backed by Zarr and Dask so you load chunks of data. Do we really want to give up on that?
Actually I think the second option is the best, only import xarray in Velocity, and remove it from the requirements altogether.

**Comment by martinRenou** [comment]
I am not saying we reimplement what xarray offers.

> Say you only want to show a subset of the data in ipyleaflet, you can select the lat/lon range xarray. If it's big data, it can be backed by Zarr and Dask so you load chunks of data. Do we really want to give up on that?

I actually don't see why it would be "giving up on that". The user could still perform whatever they wants with their data, selecting/filtering with Pandas or xarray or another library, but they would need to turn it into a simple data structure like a NumPy array before passing it to the `VelocityLayer`.
This way we don't make the choice for the user of what kind of library can be used, and we make our API simpler.

**Comment by davidbrochart** [comment]
I don't know, I also see some nice potential ipyleaflet/xarray integration. It's not implemented yet, but we could imagine ipyleaflet displaying only the lat/lon box that is currently shown, and querying the data it needs through the `xarray.Dataset`, that could trigger reading from disk or remote network... It would be similar to Datashader in that regard.

**Comment by martinRenou** [comment]
Well that is technically already possible writing some Python code. And it sounds out of scope to me for a Python port of LeafletJS for Jupyter.

**Comment by davidbrochart** [comment]
I don't know if it's out of scope, we may also want to leverage the Python ecosystem and allow the user to do powerful things while writing less code.

**Comment by martinRenou** [comment]
But that sounds really like ipyleaflet taking another route, that is what I mean. Maybe it would be better to make your idea another library.

Otherwise I fear ipyleaflet might become a huge beast with lots of dependencies that tries to fix the world's problems, but in the end doesn't do it well.
IMO it's better to have a narrow scope for libraries. A good library, in my definition, is a library that doesn't do much, but does it well.

**Comment by davidbrochart** [comment]
Yes, you are right. I think an xarray/leaflet integration deserves a separate package. This is something that has been mentioned by the xarray community (see https://github.com/benbovy/xvelmap/issues/5#issuecomment-396852106). I am going to start `xarray-leaflet` for that purpose!

**Comment by stephaniegott** [comment]
While the xarray/leaflet integration is in progress it seems like there are two options for removing the dependency on `xarray`: 
1. Use `extras_require` so that users need to install `ipyleaflet[velocity]` if they want to use the Velocity layer. 
2. Only import `xarray` in velocity (as @davidbrochart suggested) and throw an error there if users don't have it installed. 
Do either of these sound like something you folks would be okay with for right now?

**Comment by davidbrochart** [comment]
I prefer option 2. And option 1 would work with pip but not conda.

**Comment by martinRenou** [comment]
Good, should we merge this? We need to remember to remove the xarray dependency in the conda package as well when doing the release.

File: ipyleaflet/leaflet.py [hunk 1]
```diff
@@ -329,6 +329,15 @@ class VideoOverlay(RasterLayer):
 
 
 class Velocity(Layer):
+    try:
+        import xarray
+    except ModuleNotFoundError:
+        raise ImportError(
+            'The `xarray` module (required for use of the Velocity layer) is'
+            'not installed. Run "$ pip install xarray" to install the latest'
+            'version.'
```
Comment (davidbrochart) [inline ipyleaflet/leaflet.py:338]: Should we tell how to install? pip is not the only way.

Comment (stephaniegott) [reply ipyleaflet/leaflet.py]: I added the instruction for installing with conda as well.

Comment (davidbrochart) [reply ipyleaflet/leaflet.py]: I meant, maybe we should not say how to install. Sorry if it was not clear.

Comment (stephaniegott) [reply ipyleaflet/leaflet.py]: Oh I see. I do think having the instructions might be nice for someone new to these packages but I could remove them to make the message simpler.

Comment (davidbrochart) [reply ipyleaflet/leaflet.py]: I'd rather not have them. @martinRenou what do you think?
(and you don't need dask, netCDF4 or bottleneck in order to install xarray)

Comment (martinRenou) [reply ipyleaflet/leaflet.py]: I agree. Although for beginners it's nice to have an install instruction. But we should not take any guess on what kind of tool they use to install their packages, and telling someone to install something with `pip` when they actually use `conda` might break their environment if they are not careful.

File: ipyleaflet/leaflet.py [hunk 2]
```diff
@@ -334,9 +334,7 @@ class Velocity(Layer):
     except ModuleNotFoundError:
         raise ImportError(
             'The `xarray` module (required for use of the Velocity layer) is'
```
Comment (davidbrochart) [inline ipyleaflet/leaflet.py:336]: Don't we need a space after "is"?

**Comment by davidbrochart** [comment]
LGTM. Thanks @stephaniegott !

**Comment by martinRenou** [comment]
Thanks a lot!

**Comment by jasongrout** [comment]
Corresponding conda-forge issue at https://github.com/conda-forge/ipyleaflet-feedstock/issues/80

**Comment by martinRenou** [comment]
Great thanks a lot Jason :)

**Pull request merged**